---
title: "Annotation of detected homologies"
output:
  html_document: default
  pdf_document: default
date: "2023-05-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## ANNOVAR 

Annotate the origins of detected insertion using ANNOVAR (ANNOVAR Version: 2017-07-17; for more information check https://annovar.openbioinformatics.org/en/latest/user-guide/gene/). Genomic position of the first nucleotide is used.

```{python, eval = FALSE}
table_annovar.pl NAR_homologies_for_ANNOVAR.txt ~/limcr-ngs/databases/hg19/annovar/humandb/ -buildver hg19 -out NAR_homologies_for_ANNOVAR.out -protocol refGene -operation g -otherinfo –remove
```

Input file:

```{r}
NAR_homologies_for_ANNOVAR <- read.table("NAR_homologies_for_ANNOVAR.txt", header = F, sep = "\t")
head(NAR_homologies_for_ANNOVAR)
```

Load output file in R: 

```{r}
homologies_ann <- read.table("2023_05_08_NAR_homologies_ANNOVAR", header = T, sep = "\t")
head(homologies_ann)
```

## Genomic regions

```{r}
library(dplyr)
library(ggplot2)
```

Adjust sample ID. 

```{r, tidy=TRUE}

homologies_ann$sample_id <- homologies_ann$read_id
homologies_ann$sample_id <- gsub("3_NAII_","",homologies_ann$sample_id)
homologies_ann$sample_id <- gsub("_[0-9]*","",homologies_ann$sample_id)
homologies_ann$sample_id <- gsub("2NAI","",homologies_ann$sample_id)
homologies_ann$sample_id <- gsub("NAIIB","",homologies_ann$sample_id)

```

Define groups for comparison and subdivide your data frame accordingly.

```{r, tidy=TRUE}
dNTP_low <- c("HEKwtDMSO","HEKwtN","HeLaDMSO", "K11A", "K484", "wt")
dNTP_high <- c("HEKwtdN","HeLadN","K312A","N")

homologies_ann_low_dN <- homologies_ann[which(homologies_ann$sample_id %in% dNTP_low),]
homologies_ann_high_dN <- homologies_ann[which(homologies_ann$sample_id %in% dNTP_high),]
```

Create pie charts illustrating proportions of genomic regions integrated.

```{r, fig.height = 8, fig.width = 10, tidy=TRUE}
pie_labels_high <- paste0(names(table(homologies_ann_high_dN$Func.refGene)), "; ", round(100 *     table(homologies_ann_high_dN$Func.refGene)/sum(table(homologies_ann_high_dN$Func.refGene)), 2), "%")
pie(table(homologies_ann_high_dN$Func.refGene), labels = pie_labels_high)

pie_labels_low <- paste0(names(table(homologies_ann_low_dN$Func.refGene)), "; ", round(100 * table(homologies_ann_low_dN$Func.refGene)/sum(table(homologies_ann_low_dN$Func.refGene)), 2), "%")
pie(table(homologies_ann_low_dN$Func.refGene), labels = pie_labels_low)

```

## Genes

Now we want to generate a bar plot showing top100 genes affected in dNTPhigh group in a comparison to dNTPlow group.
To do so, sort the affected genes by their abundances and select top 100 genes in dNTPhigh group

```{r}
bars_low<-as.data.frame(sort(table(homologies_ann_low_dN$Gene.refGene), decreasing = TRUE))
bars_low$status<-"dNTP_low"
bars_high<-as.data.frame(sort(table(homologies_ann_high_dN$Gene.refGene), decreasing = TRUE))
bars_high$status<-"dNTP_high"
bars_high_top100 <- head(bars_high, n=100)
```

For dNTPlow group, select only the genes present in top 100 of dNTPhigh group. For aesthetics reason, if several genes are annotated for the given position, use only the first gene name for the bar plot.

```{r}
bars_low_sel<-bars_low[which(bars_low$Var1 %in% bars_high_top100$Var1),]


bars_comp<-rbind(bars_high_top100,bars_low_sel)
bars_comp$Var1 <- gsub(";.*","",bars_comp$Var1)

```

Generate the bar plot.

```{r, fig.height = 8, fig.width = 12}
ggplot(bars_comp, aes(fill = status, x = reorder(Var1, -Freq), y = Freq)) + 
  geom_bar(position="dodge", stat="identity") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
        panel.background = element_rect(fill = "white"), 
        axis.line = element_line(color="black")) + 
  labs(x = NULL , y = "Gene count")
```

## Fragile Sites

Genomic positions of detected homologies are now compared to the list of aphidicolin-sensitive genes generated by Crosetto et al. [Crosetto,N., Mitra,A., Silva,M.J., Bienko,M., Dojer,N., Wang,Q., Karaca,E., Chiarle,R., Skrzypczak,M., Ginalski,K. et al. (2013) Nucleotide-resolution DNA double-strand break mapping by next-generation sequencing. Nat. Methods, 10, 361–365., which can be downloaded from the BLESS supporting website http://breakome.utmb.edu/Home.html.


```{r}
library("openxlsx")
library("GenomicRanges")
library("IRanges")
library("dplyr")

'%ni%' <- Negate('%in%')
```

```{r}
genes<-read.table("List_of_aphidicolin_sensitive_human_genes", header = T)
head(genes)
```

For the anaylsis, use the detected homologies, that map to human genome.
```{r}
homologies_all<-read.table("2021-09-29_filtered_reads_NAR.txt",header=TRUE)
plasmids <- c("K11A","K312A","K484T", "pVT-Cas9", "wt", "d4", "SV40")
homologies <- homologies_all[which(homologies_all$sseqid %ni% plasmids),]
head(homologies)
```

Define the homology orientation on the reference.
```{r}
homologies$strand<-"+"
homologies$strand[which(homologies$sstart > homologies$send)]<-"-"
```
B
uild a data frame with ranges (= homology positions) you want to compare to the fragile sites list
```{r}
homologies_plus<-cbind(as.vector(homologies$sseqid[which(homologies$strand=="+")]), homologies$sstart[which(homologies$strand=="+")],homologies$send[which(homologies$strand=="+")],as.vector(homologies$qseqid[which(homologies$strand=="+")]),as.vector(homologies$strand[which(homologies$strand=="+")]))
colnames(homologies_plus)<-c("chr","start", "end", "id","strand")

homologies_minus<-cbind(as.vector(homologies$sseqid[which(homologies$strand=="-")]),homologies$send[which(homologies$strand=="-")],homologies$sstart[which(homologies$strand=="-")],as.vector(homologies$qseqid[which(homologies$strand=="-")]),as.vector(homologies$strand[which(homologies$strand=="-")]))
colnames(homologies_minus)<-c("chr","start", "end", "id","strand")

homologies_ranges<-data.frame(rbind(homologies_plus,homologies_minus))
head(homologies_ranges)
```



Generate first IRanges Objects and then GRanges Objects for your homology list and the sensitive genes list. Add all the desired metadata into GRanges Objects using separate columns.

```{r}
ir<-IRanges(as.integer(as.character(homologies_ranges$start)), 
            as.integer(as.character(homologies_ranges$end)),
            names=homologies_ranges$chr,
            Strang=homologies_ranges$Strang)

ir_genes <- IRanges(start=genes$start, 
                    end=genes$end,
                    names = genes$chrnum)

gr<-GRanges(seqnames=Rle(homologies_ranges$chr),
            ranges=ir,
            strand=rep(homologies_ranges$Strang),
            id=homologies_ranges$id)
gr$hom_position <- paste0(gr@seqnames,"_",gr@ranges)
names(gr) <- NULL
gr

gr_genes <- GRanges(seqnames=Rle(names(ir_genes)),
                    ranges=ir_genes, 
                    strand="+", 
                    genes=genes$geneName, 
                    cancer=genes$Cancer.Not.Cancer, 
                    geneQval = genes$qval)
gr_genes
```


Detect the overlapping regions in your target and gene reference list. This returns only a table with matching positions. 
```{r}
hits <- findOverlaps(gr, gr_genes)
head(hits)
```


To keep all the input information, filter your homology GRanges for those featuring an overlap with a  sensitive region.
```{r}
homologies_fragile_sites_OL <- subsetByOverlaps(gr, gr_genes)
homologies_fragile_sites_OL
```



To include the metadata from the gene list, extract the required data from the GRanges Object and fill the metadata columns of the Object containing the overlapping regions.

```{r}
genes_meta <- CharacterList(split(gr_genes$genes[subjectHits(hits)],
                                  queryHits(hits)))
cancer_meta <- CharacterList(split(gr_genes$cancer[subjectHits(hits)],
                                   queryHits(hits)))
qval_meta <- CharacterList(split(gr_genes$geneQval[subjectHits(hits)],
                                 queryHits(hits)))

mcols(homologies_fragile_sites_OL) <- DataFrame(mcols(homologies_fragile_sites_OL), genes_meta, cancer_meta, qval_meta)

homologies_fragile_sites_OL

```

For visualization, format the Object as data frame.
```{r}
final_fragile <- as.data.frame(homologies_fragile_sites_OL)
final_fragile$qval_meta <- as.character(final_fragile$qval_meta)
final_fragile$genes_meta <- as.character(final_fragile$genes_meta)
final_fragile$cancer_meta <- as.character(final_fragile$cancer_meta)
```

Filter for singnifcantly sensitive genes (Qval < 0.1).
Adjust sample names. 

```{r, warning=FALSE}
final_fragile$qval_meta <- as.numeric(sub(",",".", final_fragile$qval_meta))
final_fragile_sign <- final_fragile[which(final_fragile$qval_meta < 0.15),]
glimpse(final_fragile_sign)

final_fragile_sign$sample <- final_fragile_sign$id
final_fragile_sign$sample <- sub("X2_NA_I_","",final_fragile_sign$sample)
final_fragile_sign$sample <- sub("X3_NAII_","",final_fragile_sign$sample)
final_fragile_sign$sample <- sub("_[0-9].*","",final_fragile_sign$sample)
```

Devide the samples into group you want to compare and prepare your data for ggplot.
```{r}
dNTP_low <- c("HEKwt_DMSO","HEKwt_N","HeLa_DMSO", "K11A", "K484", "wt")
dNTP_high <- c("HEKwt_dN","HeLa_dN","K312A","N")

final_fragile_sign_low_dN <- final_fragile_sign[which(final_fragile_sign$sample %in% dNTP_low),]
final_fragile_sign_low_dN <- as.data.frame(table(final_fragile_sign_low_dN$genes_meta))
final_fragile_sign_low_dN$group <- "dNTP_low"
final_fragile_sign_high_dN <- final_fragile_sign[which(final_fragile_sign$sample %in% dNTP_high),]
final_fragile_sign_high_dN <- as.data.frame(table(final_fragile_sign_high_dN$genes_meta))
final_fragile_sign_high_dN$group <- "dNTP_high"


summary_fragile <- rbind(final_fragile_sign_high_dN,final_fragile_sign_low_dN)
colnames(summary_fragile) <- c("Genes", "Frequency", "Group")
```

Generate a stacked plot, presenting the integration frequencies from different fragile genes within the defined groups.

```{r}
ggplot(summary_fragile, aes(fill=Genes, y=Frequency, x=Group)) + 
  geom_bar(position="stack", stat="identity")

```

















